### PR TITLE
Short link info and generic goldenlayout

### PR DESCRIPTION
**Pull request by partouf**
For the mobile app I'm making, I wondered if I could get it to open CE shortlinks, but shortlink is stored in its original goldenlayout format, which has way way way more info than I would like to receive.

So the first thing this PR adds is:
* /api/shortlinkinfo/:id - translates goldenlayout to basic client state and returns in JSON format

And for fun, because Matt was talking about being able to store state without having to depend on GL, I made:
* /resetlayout/:id - rebuilds a new goldenlayout from the basic client state that was derived from the original goldenlayout, so same code and compilers, different clean layout (so this is able to fix broken /z/ links)

Offers a workaround for https://github.com/mattgodbolt/compiler-explorer/issues/1077 and other goldenlayout issues.

**Todo now**
- [x] API for Apps to load clientstate from links
- [x] Better name for /y/ -> now is /resetlayout/
- [x] Option to make shortlinks from clientstate

**Todo later**
- Add optional layout to clientstate
- Able to interpret goldenlayout to clientstate layout
- Able to regenerate goldenlayout from clientstate layout

### CODE REVIEW

**Comment by mattgodbolt** [comment]
Thanks @partouf ! Will take a look at this tomorrow!

File: lib/clientstate-normalizer.js [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+                {
+                    type: "component",
+                    componentName: "opt",
+                    componentState: {
+                        id: compilerIndex,
+                        editorid: session.id
+                    },
+                    isClosable: true,
+                    reorderEnabled: true,
+                    title: compiler.id + " Opt Viewer (Editor #" + session.id + ", Compiler #" + compilerIndex
```
[the remark this answers is not in the document]
Comment (RabsRincon) [reply lib/clientstate-normalizer.js]: The title is regenerated clientside using the ids on its state, so you're right, there should be no need to duplicate it here

File: app.js [hunk 2]
```diff
@@ -483,6 +571,8 @@ aws.initConfig(awsProps)
                     .use('/api', apiHandler.handle)
                     .use('/g', oldGoogleUrlHandler)
                     .get('/z/:id', storedStateHandler)
+                    .get('/y/:id', storedStateHandlerRebuilt)
```
[the remark this answers is not in the document]
Comment (partouf) [reply app.js]: agree with shortlinkinfo, should be in api, was planning on doing that but lazy/forgot

File: lib/clientstate-normalizer.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (partouf) [reply lib/clientstate-normalizer.js]: yeah the title is regenerated, it was really only for debugging, ill remove it

File: lib/clientstate-normalizer.js [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+                    reorderEnabled: true,
+                    title: compiler.id + " (Editor #" + session.id + ") " + session.language
+                }
+            ]
+        };
+    }
+
+    fromClientState(state) {
+        this.golden = {
+            settings: {
```
[the remark this answers is not in the document]
Comment (partouf) [reply lib/clientstate-normalizer.js]: I just copy/pasted a freshly reset state, and then it worked, so I have no idea what it all means

File: lib/clientstate.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ,
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+"use strict";
+
+class ClientStateCompilerOptions {
```
[the remark this answers is not in the document]
Comment (partouf) [reply lib/clientstate.js]: It's true that there's no layout information. I had some ideas, but not entirely sure yet what the best way is to approach that.

**Comment by partouf** [comment]
Finally figured out what to do with the /y/:id link. I renamed it to /resetlayout/:id and added an option in the dropdown menu for it when you're at a /z/ link page.

Reset link

**Pull request merged**